consensus/src/decoupled_state_computer.rs

```diff
@@ -0,0 +1,128 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{error::StateSyncError, state_replication::StateComputer};
+use anyhow::Result;
+use consensus_types::{block::Block, executed_block::ExecutedBlock};
+use diem_crypto::{HashValue, hash::ACCUMULATOR_PLACEHOLDER_HASH};
+use diem_infallible::Mutex;
+use diem_logger::prelude::*;
+use diem_metrics::monitor;
+use diem_types::ledger_info::LedgerInfoWithSignatures;
+use execution_correctness::ExecutionCorrectness;
+use executor_types::{Error as ExecutionError, StateComputeResult};
+use fail::fail_point;
+use state_sync::client::StateSyncClient;
+use std::{boxed::Box, sync::Arc};
+
+/// Basic communication with the Execution module;
+/// implements StateComputer traits.
+pub struct DecoupledExecutionProxy {
+    execution_correctness_client: Mutex<Box<dyn ExecutionCorrectness + Send + Sync>>,
+    synchronizer: StateSyncClient,
+}
+
+impl DecoupledExecutionProxy {
+    pub fn new(
+        execution_correctness_client: Box<dyn ExecutionCorrectness + Send + Sync>,
+        synchronizer: StateSyncClient,
+    ) -> Self {
+        Self {
+            execution_correctness_client: Mutex::new(execution_correctness_client),
+            synchronizer,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl StateComputer for DecoupledExecutionProxy {
+    fn compute(
+        &self,
+        // The block to be executed.
+        block: &Block,
+        // The parent block id.
+        parent_block_id: HashValue,
+    ) -> Result<StateComputeResult, ExecutionError> {
+        fail_point!("consensus::compute", |_| {
+            Err(ExecutionError::InternalError {
+                error: "Injected error in compute".into(),
+            })
+        });
+        debug!(
+            block_id = block.id(),
+            parent_id = block.parent_id(),
+            "Executing block",
+        );
+
+        // Return dummy block and bypass the execution phase.
+        // This will break the e2e smoke test.
+        monitor!(
+            "execute_block",
+            Ok(StateComputeResult::new(
+                *ACCUMULATOR_PLACEHOLDER_HASH,
+                vec![],
+                0,
+                vec![],
+                0,
+                None,
+                vec![],
+                vec![],
+                vec![],
+            ))
+        )
+
+    }
+
+    /// Send a successful commit. A future is fulfilled when the state is finalized.
+    async fn commit(
+        &self,
+        blocks: &[Arc<ExecutedBlock>],
+        finality_proof: LedgerInfoWithSignatures,
+    ) -> Result<(), ExecutionError> {
+        let mut block_ids = Vec::new();
```

Comment:
let's just use `unimplemented()` instead of copy-paste the code

Reply:
Resolved